h2o-core/src/main/java/water/fvec/NewChunk.java

```diff
@@ -567,12 +568,29 @@ private void append_ss(BufferedString str) {
     byte b[] = str.getBuffer();
 
     if (_ss == null) {
-      _ss = MemoryManager.malloc1((strlen + 1) * 4);
-    }
-    while (_ss.length < (_sslen + strlen + 1)) {
-      _ss = MemoryManager.arrayCopyOf(_ss,_ss.length << 1);
+      int allocSize = (strlen + 1) * 4;
+      if (allocSize < 0 || allocSize > MAX_ARRAY_SIZE) {
+        allocSize = MAX_ARRAY_SIZE;
+        strlen = Math.min(strlen, MAX_ARRAY_SIZE);
+      }
+      _ss = MemoryManager.malloc1(Math.min(Math.abs(allocSize), MAX_ARRAY_SIZE));
+    }
+    int sizeBoundary = _sslen + strlen + 1;
+    while (_ss.length < sizeBoundary) {
+      // An attempt to double the capacity
+      int newSize = _ss.length << 1;
+      if (newSize < 0 || newSize > MAX_ARRAY_SIZE) {
+        newSize = MAX_ARRAY_SIZE;
+        _ss = MemoryManager.arrayCopyOf(_ss, newSize);
+        if (strlen > newSize) strlen = sizeBoundary - newSize;
+        break;
+      } else {
+        _ss = MemoryManager.arrayCopyOf(_ss, newSize);
+      }
     }
-    for (int i = off; i < off+strlen; i++)
+
+
+    for (int i = off; i < off + strlen - 1; i++)
       _ss[_sslen++] = b[i];
     _ss[_sslen++] = (byte)0; // for trailing 0;
```

Comment:
The code causes an overflow of array index here in testHugeChunkExpand (at index 2147483639). Have to figure out why the index shift is wrong.